### ISSUE TITLE
Ability to calculate projected memory usage for a given model

### ISSUE THREAD

**Issue by Bidski** [tensorflow/tensorflow#36327]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.1
- Are you willing to contribute it (Yes/No): Yes, but I would likely need assistance/guidance

**Describe the feature and the current behavior/state.**
The closest feature I can find to this is `tf.keras.Model.summary()`. This feature calculate the trainable and non-trainable parameters per-layer but doesn't attempt to calculate any memory usage statistics.

The proposed feature would extend `tf.keras.Model.summary()` by also calculating memory requirements per-layer and for the complete model. If multiple compute devices are used in the model then a per-device memory breakdown would also be useful.

**Will this change the current api? How?**
This could either just add functionality to `tf.keras.Model.summary()` without changing the interface, or it could add extra parameters to `tf.keras.Model.summary()`. Alternatively, a new function could be added, whatever is deemed most appropriate.

**Who will benefit with this feature?**
Anyone with a desire to get an idea about how much memory their model will need for either training or inference. It should be possible for someone to determine whether the model they wish to train/evaluate will fit into RAM (either system of GPU) that they have available.

**Any Other info.**
My focus is on the `tf.keras.Model.summary()` API, but this feature should probably be extended to other means of model creation.

**Comment by jamesmishra** [comment]
The answers to [this StackOverflow question](https://stackoverflow.com/questions/43137288/how-to-determine-needed-memory-of-keras-model/) have some (approximate?) functions for estimating the GPU memory usage of a Keras model.

I contributed an answer to the question, which is:
```python
def keras_model_memory_usage_in_bytes(model, *, batch_size: int):
    """
    Return the estimated memory usage of a given Keras model in bytes.
    This includes the model weights and layers, but excludes the dataset.

    The model shapes are multipled by the batch size, but the weights are not.

    Args:
        model: A Keras model.
        batch_size: The batch size you intend to run the model with. If you
            have already specified the batch size in the model itself, then
            pass `1` as the argument here.
    Returns:
        An estimate of the Keras model's memory usage in bytes.

    """
    default_dtype = tf.keras.backend.floatx()
    shapes_mem_count = 0
    internal_model_mem_count = 0
    for layer in model.layers:
        if isinstance(layer, tf.keras.Model):
            internal_model_mem_count += keras_model_memory_usage_in_bytes(
                layer, batch_size=batch_size
            )
        single_layer_mem = tf.as_dtype(layer.dtype or default_dtype).size
        out_shape = layer.output_shape
        if isinstance(out_shape, list):
            out_shape = out_shape[0]
        for s in out_shape:
            if s is None:
                continue
            single_layer_mem *= s
        shapes_mem_count += single_layer_mem

    trainable_count = sum(
        [tf.keras.backend.count_params(p) for p in model.trainable_weights]
    )
    non_trainable_count = sum(
        [tf.keras.backend.count_params(p) for p in model.non_trainable_weights]
    )

    total_memory = (
        batch_size * shapes_mem_count
        + internal_model_mem_count
        + trainable_count
        + non_trainable_count
    )
    return total_memory

```

**Comment by Bidski** [comment]
@jamesmishra I have found that this method vastly underestimates the amount of GPU memory needed by my models (this method reports ~5GB, however my models with a batch size of 1 are unable to train on a 32GB V100).

**Comment by jamesmishra** [comment]
@Bidski I've noticed around a 5% - 10% discrepancy on my own models, but they are comparatively smaller models. If you're willing to send me your `model.summary()` and the `input_shape`, I'ld to see if I can replicate the issue you're seeing. My email is j@jamesmishra.com

I don't want to clutter up the TensorFlow issue tracker, so I'll post further discussion and updates [on this GitHub Gist](https://gist.github.com/jamesmishra/34bac09176bc07b1f0c33886e4b19dc7).

**Comment by reidswanson** [comment]
Hi, I've been interested in this as well and still haven't found an adequate solution. I tried using the method suggested by @jamesmishra, but I can't get it to work (the output_shape is always `None` even if I compile and call the model on dummy data).

In any case, I don't believe your method is likely to be robust across model architectures, although this is not my area of expertise at all. [Microsoft has a recent paper](https://www.microsoft.com/en-us/research/publication/estimating-gpu-memory-consumption-of-deep-learning-models/) on trying to estimate the required memory for a model and it is significantly more complex. Unfortunately, their code is not published (at least anywhere I can find). It would be really nice to have something along the lines of what they describe built into the toolkit.

**Comment by jamesmishra** [comment]
I'm always looking for feedback for my code snippet. If you have anything, feel free to comment on [this GitHub gist](https://gist.github.com/jamesmishra/34bac09176bc07b1f0c33886e4b19dc7).

It would be great to get Microsoft to open-source their code, but the good news is that most of us would be happier with a much simpler function that only covers the memory usage of the core Keras layers. That can't be that hard.

**Comment by reidswanson** [comment]
@jamesmishra I agree it would still be great to have something that works for the basic Keras layers and 80% of use cases and I'm very appreciative of yours (and others) work on the problem. 

It's still frustrating though that for the models I'm working on (GPT like transformers), and apparently other common models, the simpler approaches can wildly underestimate the memory needed depending on the hyper-parameters of the model. If I gain any insights why or when I'll be sure to comment on your gist with my thoughts and discoveries.

**Comment by zainulabidin302** [comment]
A work around could to use for example 
```
model_object = SomeSubclassedModel()
inputs = tf.keras.layers.Input(shape=(256, 512, 3))
sequential = tf.keras.Model(inputs=inputs, output=model_object(inputs))
sequential.summary() # i think now the memory can be better estimated because all shapes are computed.
```

To incorporate this in a class one could do for example

```
class SomeSubclassedModel(keras.Model): 
   ...
   def call(self, x): 
       ...
   def get_seq_model(self):
      inputs = tf.keras.layers.Input(shape=(256, 512, 3))
      tf.keras.Model(inputs=inputs, output=self.call(inputs))
      return tf.keras.Model(input=inputs, output=self.call(inputs))

```

To include batch_size estimation too, we could instantiate the model with random normal distribution for a trail run.

```
class SomeSubclassedModel(keras.Model): 
   ...
   def call(self, x): 
       ...
   def get_seq_model(self, shape=()):
      inputs = tf.random.normal(full_shape)
      model = tf.keras.Model(input=inputs, output=self.call(inputs))
      # monkey patching 
      def keras_model_memory_usage_in_bytes():
           ....
      model.memory_usage = memory_usage
      return 

```

Later use it 

```
model = SomeSubclassedModel()
model.get_seq_model(shape=(batch_size, H, W, C)).summary()
model.get_seq_model(). keras_model_memory_usage_in_bytes()
```

Once we use `call` on a subclasses object, all shapes are computed. This could be a source of information for computing the memory. What else would be required to compute the complete memory profile? I am not sure.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by Bidski** [comment]
I do not see anything in the release notes or the `tf.keras.Model.summary()` API docs that looks like it addresses anything mentioned in this feature request.

**Comment by Venkat6871** [comment]
Hi @Bidski ,
Thank you so much for filing this feature request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!
